File: python/src/ppg_hr/core/fft_peaks.py
"""Detect spectral peaks within the cardiac band (port of ``FFT_Peaks.m``).

Pads ``signal`` to ``2**13`` samples, computes the one-sided amplitude
spectrum, then returns peaks whose amplitude exceeds ``percent`` of the local
maximum AND whose 1-based bin index lies strictly inside the open interval
``(1*Len/Fs + 1, 4*Len/Fs)`` (i.e. the cardiac 1–4 Hz window in bin indices).
"""

from __future__ import annotations

import numpy as np
from scipy.signal import find_peaks

__all__ = ["fft_peaks"]

_FFT_LEN: int = 1 << 13  # 8192


def fft_peaks(
    signal: np.ndarray, fs: float, percent: float
) -> tuple[np.ndarray, np.ndarray]:
    """Return ``(frequencies_hz, amplitudes)`` of cardiac-band spectral peaks."""
    sig = np.atleast_1d(np.asarray(signal, dtype=float)).ravel()
    a = sig.size
    if a == 0:
        return np.array([]), np.array([])

    spectrum = np.fft.fft(sig, _FFT_LEN)
    amp_full = np.abs(spectrum) / a
    half = _FFT_LEN // 2
    amp1 = amp_full[:half].copy()
    amp1[1:] *= 2.0
    freq_axis = fs * np.arange(half) / _FFT_LEN

    # ``locs`` from MATLAB ``findpeaks`` is 1-based; scipy.signal.find_peaks
    # returns 0-based indices, so add 1 when comparing with the original
    # band edges to preserve MATLAB semantics exactly.
    peaks_idx, _ = find_peaks(amp1)
    if peaks_idx.size == 0:
        return np.array([]), np.array([])

    locs_1based = peaks_idx + 1
    free_low = 1.0 * _FFT_LEN / fs + 1.0
    free_high = 4.0 * _FFT_LEN / fs

    valid_mask = (locs_1based < free_high) & (locs_1based > free_low)
    pks2 = amp1[peaks_idx][valid_mask]
    if pks2.size == 0:
        return np.array([]), np.array([])

    threshold = pks2.max() * percent
    keep_mask = (amp1[peaks_idx] > threshold) & valid_mask
    keep_idx = peaks_idx[keep_mask]
    return freq_axis[keep_idx], amp1[keep_idx]

File: MATLAB/FFT_Peaks.m
% function [Fre,Famp] = FFT_Peaks(Signal,Fs,percent)  
% %输出大于百分之多少的maxPeak幅值的谱峰位置和幅值,峰必须在1~3Hz
% %Signal 要做fft的信号
% %Fs 采样率
% %percent 需要提取的大于最大peak幅值百分之多少的peaks
% 
% a = max(size(Signal)); %信号长度
% Len = 2^13;            %补零后的长度
% if size(Signal,1) > size(Signal,2)
%     Signal = Signal';
% end
% 
% %加窗  看起来泄露现象好一点 不会有明显的旁瓣
% % w = hann(length(Signal))'; %汉宁窗主瓣够窄，比较好。
% % Signal = Signal.*w;
% 
% FFTData = fft(Signal, Len);
% FFTAmp0 = abs(FFTData)/a;                 %频谱关于Len/2对称
% FFTAmp1 = FFTAmp0(1:Len/2); 
% FFTAmp1(2:end) = 2*FFTAmp1(2:end);
% Frequence1 = Fs*((0:(Len/2)-1))/Len;
% 
% free_low = 1*Len/Fs+1;
% free_high = 4*Len/Fs;  %峰必须在1~3Hz之间
% [pks,locs] = findpeaks(FFTAmp1);  %找所有的峰
% pks_2 = pks((locs < free_high) & (locs>free_low));
% locss = locs((pks>max(pks_2)*percent));
% locss = locss((locss < free_high) & (locss>free_low));  
% Fre = Frequence1(locss);
% Famp = FFTAmp1(locss);
% 
% rmss = rms(FFTAmp1);
% ppeak = max(FFTAmp1);
% cff = ppeak/rmss;
% end

function [Fre, Famp] = FFT_Peaks(Signal, Fs, percent)
    % 确保输入是行向量
    if size(Signal,1) > size(Signal,2)
        Signal = Signal';
    end

    a = length(Signal); 
    Len = 2^13;            
    
    FFTData = fft(Signal, Len);
    FFTAmp0 = abs(FFTData)/a;                 
    FFTAmp1 = FFTAmp0(1:Len/2); 
    FFTAmp1(2:end) = 2*FFTAmp1(2:end);
    Frequence1 = Fs*((0:(Len/2)-1))/Len;

    % 定义有效频率范围 (通常为 1~4 Hz)
    free_low = 0.7 * Len/Fs + 1;
    free_high = 4 * Len/Fs; 
    
    [pks, locs] = findpeaks(FFTAmp1); 
    
    % --- 修复开始 ---
    % 筛选出在有效频率范围内的峰
    valid_indices = (locs < free_high) & (locs > free_low);
    pks_2 = pks(valid_indices);
    
    % 如果有效范围内没有峰 (例如静止状态或信号太弱)
    if isempty(pks_2)
        Fre = []; 
        Famp = [];
        return; % 直接返回空值，防止下方 max() 报错
    end
    % --- 修复结束 ---

    % 只有在 pks_2 非空时，才计算阈值
    threshold = max(pks_2) * percent;
    
    % 再次筛选：必须大于阈值，且在有效范围内
    locss = locs((pks > threshold) & valid_indices);
    
    Fre = Frequence1(locss);
    Famp = FFTAmp1(locss);
end
